chain/postgresdb/database/database.go

```diff
@@ -0,0 +1,342 @@
+		}
+		l.Infow("rollback tran")
+	}()
+
+	if err := fn(tx); err != nil {
+		//nolint:errorlint // Have no clue why this is a problem :(.
+		if pqerr, ok := err.(*pq.Error); ok && pqerr.Code == uniqueViolation {
+			return ErrDBDuplicatedEntry
+		}
+		return fmt.Errorf("exec tran: %w", err)
+	}
+
+	if err := tx.Commit(); err != nil {
+		return fmt.Errorf("commit tran: %w", err)
+	}
+	l.Infow("commit tran")
+
+	return nil
+}
+
+// ExecContext is a helper function to execute a CUD operation with
+// logging and tracing.
+func ExecContext(ctx context.Context, l log.Logger, db sqlx.ExtContext, query string) error {
+	return NamedExecContext(ctx, l, db, query, struct{}{})
+}
+
+// NamedExecContext is a helper function to execute a CUD operation with
+// logging and tracing where field replacement is necessary.
+func NamedExecContext(ctx context.Context, l log.Logger, db sqlx.ExtContext, query string, data any) error {
+	q, err := queryString(query, data)
+	if err != nil {
+		return err
+	}
+
+	if _, ok := data.(struct{}); ok {
+		//nolint:gomnd // Having to add constants is overkill.
+		l.AddCallerSkip(2).Infow("database.NamedExecContext", "query", q)
+	} else {
+		l.AddCallerSkip(1).Infow("database.NamedExecContext", "query", q)
+	}
+
+	if _, err := sqlx.NamedExecContext(ctx, db, query, data); err != nil {
+		//nolint:errorlint // Have no clue why this is a problem :(.
+		if pqerr, ok := err.(*pq.Error); ok {
+			switch pqerr.Code {
+			case undefinedTable:
+				return ErrUndefinedTable
+			case uniqueViolation:
+				return ErrDBDuplicatedEntry
+			}
+		}
+		return err
+	}
+
+	return nil
+}
+
+// QuerySlice is a helper function for executing queries that return a
+// collection of data to be unmarshalled into a slice.
+func QuerySlice[T any](ctx context.Context, l log.Logger, db sqlx.ExtContext, query string, dest *[]T) error {
+	return NamedQuerySlice(ctx, l, db, query, struct{}{}, dest)
+}
+
+// NamedQuerySlice is a helper function for executing queries that return a
+// collection of data to be unmarshalled into a slice where field replacement is necessary.
+func NamedQuerySlice[T any](ctx context.Context, l log.Logger, db sqlx.ExtContext, query string, data any, dest *[]T) error {
+	q, err := queryString(query, data)
+	if err != nil {
+		return err
+	}
+
+	if _, ok := data.(struct{}); ok {
+		//nolint:gomnd // Having to add constants is overkill.
+		l.AddCallerSkip(2).Infow("database.QuerySlice", "query", q)
+	} else {
+		l.AddCallerSkip(1).Infow("database.QuerySlice", "query", q)
+	}
+
+	rows, err := sqlx.NamedQueryContext(ctx, db, query, data)
+	if err != nil {
+		//nolint:errorlint // We want this
+		if pqerr, ok := err.(*pq.Error); ok && pqerr.Code == undefinedTable {
+			return ErrUndefinedTable
+		}
+		return err
+	}
+	defer rows.Close()
+
+	var slice []T
+	for rows.Next() {
+		v := new(T)
+		if err := rows.StructScan(v); err != nil {
+			return err
+		}
+		slice = append(slice, *v)
+	}
+	*dest = slice
```

Comment:
We use `*dest` here to assign to the underlying memory address referenced by the pointer rather than to the pointer itself. That way, we keep the pointer between the two functions but change what they point to. In this case, it allows us to grow the slice as needed by the results from the database.